cos_alerter/server.py

```diff
@@ -44,16 +54,32 @@ def alive():
     # TODO Decide if we should validate the request.
     params = request.args
     clientid_list = params.getlist("clientid")  # params is a werkzeug.datastructures.MultiDict
-    if len(clientid_list) < 1:
-        logger.warning("Request %s has no clientid.", request.url)
-        return 'Parameter "clientid" required.', 400
-    if len(clientid_list) > 1:
-        logger.warning("Request %s specified clientid more than once.", request.url)
-        return 'Parameter "clientid" provided more than once.', 400
+    key_list = params.getlist("key")
+
+    if len(clientid_list) < 1 or len(key_list) < 1:
+        logger.warning("Request %s is missing clientid or key.", request.url)
+        return 'Parameters "clientid" and "key" are required.', 400
+    if len(clientid_list) > 1 or len(key_list) > 1:
+        logger.warning("Request %s specified clientid or key more than once.", request.url)
+        return 'Parameters "clientid" and "key" should be provided exactly once.', 400
     clientid = clientid_list[0]
-    if clientid not in config["watch"]["clients"]:
-        logger.warning("Request %s specified an unknown clientid.")
+    key = key_list[0]
+    try:
+        uuid.UUID(clientid, version=4)
+    except ValueError:
+        logger.warning("Request %s specified an invalid clientid.", request.url)
+        return 'Clientid {params["clientid"]} is not a valid UUID. ', 400
+
+    clients = config["watch"]["clients"]
+    # Find the client with the specified clientid
+    matching_clients = [c for c in clients if c["id"] == clientid]
+    if not matching_clients:
+        logger.warning("Request %s specified an unknown clientid.", request.url)
         return 'Clientid {params["clientid"]} not found. ', 404
+    client = matching_clients[0]
+    if key != client["key"]:
+        logger.warning("Request %s provided an incorrect key.", request.url)
+        return "Incorrect key for the specified clientid.", 401
```

Comment:
We should not be storing keys directly but hashes of the keys, and comparing to the hashed value from the request.